### ISSUE TITLE
APML is broken

### ISSUE THREAD

**Issue by attacomsian** [marytts/marytts#469]
Input type: `APML`
Output type: `AUDIO`

Input Text
```xml
<?xml version="1.0" ?>
<!DOCTYPE apml SYSTEM "http://mary.dfki.de/lib/apml.dtd" []>
<apml xml:lang="en-US">
<performative  type="announce">
<theme affect="joy">
Welcome<boundary type="H"/>to the <emphasis level="strong">wonderful</emphasis> world of
<emphasis x-pitchaccent="Hstar">speech</emphasis> synthesis
<boundary type="LL"/>!
</theme>
</performative>
</apml>
```

```
Internal server error

2016-02-01 15:00:32,515 [I/O dispatcher 6] INFO  marytts.R 3 New request (input type "APML", output type "AUDIO", voice "cmu-slt-hsmm", effect "", audio "WAVE")
2016-02-01 15:00:32,515 [I/O dispatcher 6] DEBUG marytts.IO Parsing XML input (non-validating): <?xml version="1.0" ?>
<!DOCTYPE apml SYSTEM "http://mary.dfki.de/lib/apml.dtd" []>
<apml xml:lang="en-US">
<performative  type="announce">
<theme affect="joy">
Welcome<boundary type="H"/>to the <emphasis level="strong">wonderful</emphasis> world of
<emphasis x-pitchaccent="Hstar">speech</emphasis> synthesis
<boundary type="LL"/>!
</theme>
</performative>
</apml>

2016-02-01 15:00:32,522 [I/O dispatcher 6] WARN  marytts.server Problem reading input
java.io.FileNotFoundException: http://mary.dfki.de/lib/apml.dtd
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1835)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1440)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.setupCurrentEntity(XMLEntityManager.java:646)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startEntity(XMLEntityManager.java:1300)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startDTDEntity(XMLEntityManager.java:1267)
	at com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.setInputSource(XMLDTDScannerImpl.java:263)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.dispatch(XMLDocumentScannerImpl.java:1164)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.next(XMLDocumentScannerImpl.java:1050)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:938)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:606)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:117)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:510)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:848)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:777)
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:243)
	at com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:348)
	at marytts.util.dom.DomUtils.parseDocument(DomUtils.java:142)
	at marytts.util.dom.DomUtils.parseDocument(DomUtils.java:121)
	at marytts.datatypes.MaryData.setData(MaryData.java:256)
	at marytts.server.Request.setInputData(Request.java:253)
	at marytts.server.http.SynthesisRequestHandler.process(SynthesisRequestHandler.java:220)
	at marytts.server.http.SynthesisRequestHandler.handleClientRequest(SynthesisRequestHandler.java:91)
	at marytts.server.http.BaseHttpRequestHandler.handle(BaseHttpRequestHandler.java:138)
	at org.apache.http.nio.protocol.BufferingHttpServiceHandler$RequestHandlerAdaptor.handle(BufferingHttpServiceHandler.java:189)
	at org.apache.http.nio.protocol.SimpleNHttpRequestHandler.handle(SimpleNHttpRequestHandler.java:51)
	at org.apache.http.nio.protocol.AsyncNHttpServiceHandler.processRequest(AsyncNHttpServiceHandler.java:453)
	at org.apache.http.nio.protocol.AsyncNHttpServiceHandler.requestReceived(AsyncNHttpServiceHandler.java:225)
	at org.apache.http.nio.protocol.BufferingHttpServiceHandler.requestReceived(BufferingHttpServiceHandler.java:127)
	at org.apache.http.impl.nio.DefaultNHttpServerConnection.consumeInput(DefaultNHttpServerConnection.java:161)
	at org.apache.http.impl.nio.DefaultServerIOEventDispatch.inputReady(DefaultServerIOEventDispatch.java:147)
	at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:161)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:335)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:275)
	at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:542)
	at java.lang.Thread.run(Thread.java:745)
2016-02-01 15:00:32,523 [I/O dispatcher 6] DEBUG marytts.http Returning HTTP status 500: Problem reading input
java.io.FileNotFoundException: http://mary.dfki.de/lib/apml.dtd
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1835)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1440)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.setupCurrentEntity(XMLEntityManager.java:646)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startEntity(XMLEntityManager.java:1300)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startDTDEntity(XMLEntityManager.java:1267)
	at com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.setInputSource(XMLDTDScannerImpl.java:263)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.dispatch(XMLDocumentScannerImpl.java:1164)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.next(XMLDocumentScannerImpl.java:1050)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:938)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:606)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:117)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:510)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:848)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:777)
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:243)
	at com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:348)
	at marytts.util.dom.DomUtils.parseDocument(DomUtils.java:142)
	at marytts.util.dom.DomUtils.parseDocument(DomUtils.java:121)
	at marytts.datatypes.MaryData.setData(MaryData.java:256)
	at marytts.server.Request.setInputData(Request.java:253)
	at marytts.server.http.SynthesisRequestHandler.process(SynthesisRequestHandler.java:220)
	at marytts.server.http.SynthesisRequestHandler.handleClientRequest(SynthesisRequestHandler.java:91)
	at marytts.server.http.BaseHttpRequestHandler.handle(BaseHttpRequestHandler.java:138)
	at org.apache.http.nio.protocol.BufferingHttpServiceHandler$RequestHandlerAdaptor.handle(BufferingHttpServiceHandler.java:189)
	at org.apache.http.nio.protocol.SimpleNHttpRequestHandler.handle(SimpleNHttpRequestHandler.java:51)
	at org.apache.http.nio.protocol.AsyncNHttpServiceHandler.processRequest(AsyncNHttpServiceHandler.java:453)
	at org.apache.http.nio.protocol.AsyncNHttpServiceHandler.requestReceived(AsyncNHttpServiceHandler.java:225)
	at org.apache.http.nio.protocol.BufferingHttpServiceHandler.requestReceived(BufferingHttpServiceHandler.java:127)
	at org.apache.http.impl.nio.DefaultNHttpServerConnection.consumeInput(DefaultNHttpServerConnection.java:161)
	at org.apache.http.impl.nio.DefaultServerIOEventDispatch.inputReady(DefaultServerIOEventDispatch.java:147)
	at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:161)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:335)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:275)
	at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:542)
	at java.lang.Thread.run(Thread.java:745)

```

**Comment by psibre** [comment]
> `java.io.FileNotFoundException: http://mary.dfki.de/lib/apml.dtd`

Not sure why it's trying to load that. According to the [Wayback Machine](https://web.archive.org/web/*/http://mary.dfki.de/lib/*), the APML DTD doesn't seem to have ever existed there...

**Comment by psibre** [comment]
OK, going through `git log --all -- **/apml.dtd` shows that this issue was probably triggered by f23f022fe22b23d4fb326f07d365830ca2aeb1c2, five years ago.

**Comment by attacomsian** [comment]
It tries to validate the XML input with the DTD mentioned in input text: `http://mary.dfki.de/lib/apml.dtd`
This validation is controlled via property: `maryxml.validate.input` 
```
<?xml version="1.0" ?>
<!DOCTYPE apml SYSTEM "http://mary.dfki.de/lib/apml.dtd" []>
<apml xml:lang="en-US">
<performative  type="announce">
<theme affect="joy">
Welcome<boundary type="H"/>to the <emphasis level="strong">wonderful</emphasis> world of
<emphasis x-pitchaccent="Hstar">speech</emphasis> synthesis
<boundary type="LL"/>!
</theme>
</performative>
</apml>
```
Same is happening in case of `SABLE` #468. I found the DTDs and will add them into project.

**Comment by psibre** [comment]
#280 depends on this

**Comment by psibre** [comment]
From what I can see, there are two possible fixes:
## Elegant solution
I think the first part of this solution is indeed to add the APML DTD as a resource, like in your 4ee73dd3208d297d35bb5afe98cf26e5d32bf32a.
But then implement a custom [EntityResolver](http://docs.oracle.com/javase/7/docs/api/org/xml/sax/EntityResolver.html) that loads that resource, and set it to be used somewhere in DomUtils.
See e.g. [this SO answer](http://stackoverflow.com/a/4726522) for more details.
## Lazy solution
Just provide an empty DTD, as suggested [here](http://marcels-javanotes.blogspot.de/2005/11/parsing-xml-file-without-having-access.html)

**Comment by attacomsian** [comment]
Thank you for providing links for **Elegant Solution**. It was really helpful!

I created a custom class `MaryEntityResolver.java` to resolve the `APML` and `SABLE` DTDs. It keeps the DTDs resource URL same in demo texts and resolves them internally from classpath. Also, I opened new PR #496 which includes fix for this and #468.

**Comment by psibre** [comment]
Fixed by #496
